### ISSUE TITLE
Access to MPI_Barrier as an op?

### ISSUE THREAD

**Issue by ppwwyyxx** [uber/horovod#159]
My use case: I wanted to do some extra things (e.g. validation) in process rank 0, and would like other processes to wait for it.
If other processes don't wait for it, when it takes too long on process 0 to do these things, I saw things like: 
```
WARNING: One or more tensors were submitted to be reduced, gathered or broadcasted by subset
 of ranks and are waiting for remainder of ranks for more than 60 seconds. This may indicate that 
different ranks are trying to submit different tensors or that only subset of ranks is submitting 
tensors, which will cause deadlock. Stalled ops: 
HVDAllReduce/HorovodAllreduce_gradients_AddN_69_0 [ready ranks: 2, 6, 7, 3, 4, 1, 5],
```

After many of this messages the training hangs forever. I guess after they waited too long they will give up eventually?

Is it OK to add an op which calls MPI_Barrier?
Maybe I can also give a try to mpi4py, etc. Not sure if mixing two MPI libraries together is a good idea.

**Comment by alsrgv** [comment]
The purpose of this warning is to show to the user that there's a large asymmetry in his workload.  Since you have many GPUs waiting, can you shard your validation load on all of them, and then simply average metrics with allreduce, like [here](https://github.com/uber/horovod/blob/master/examples/keras_mnist_advanced.py#L91)?

There's no "give up" mechanics, so if you see these messages forever, it means that something got stuck.

Mixing with mpi4py should be fine, as long as you use multi-threaded MPI build (Open MPI 3.0.0 is multi-threaded by default) and you tell mpi4py to not re-initialize:

```python
import mpi4py
mpi4py.rc.initialize = False

# first, initialize horovod
hvd.init()

# then, use mpi4py
from mpi4py import MPI as mpi
gathered = mpi.allgather(obj)
```

**Comment by alsrgv** [comment]
@ppwwyyxx, actually, we just realized internally that we'd need to make a small change to allow other MPI operations to coexist with Horovod.  We plan to release a large refactor with a bundle of changes including this in a couple of weeks, but let us know if you want to try this earlier and I can push a private branch for you.

**Comment by ppwwyyxx** [comment]
Thanks! I do hope to have access to MPI primitives like MPI_Barrier.

I understand that keeping all other GPUs waiting is not a good idea for performance, but it can be very helpful during debugging. Also, for metrics other than accuracy, it's not always easy to shard and distribute the evaluation tasks.

**Comment by alsrgv** [comment]
@ppwwyyxx, `hvd.allreduce([], name="Barrier")` would produce an equivalent effect to an MPI_Barrier.  But it would still print a warning that other nodes are waiting on the node running the validation, which can be ignored.  Would that work for you?  We can think about finer control over warnings if you don't want to see them.

**Comment by kelsiuss** [comment]
@alsrgv Is there any way to control the warnings like you mentioned? Not sure if anything was implemented to manage this since you last posted.

**Comment by alsrgv** [comment]
@kelsiuss, not yet.  Would flag to disable warnings be useful?

**Comment by avolkov1** [comment]
This idiom for barrier works very well with Keras+TF:
```
import horovod.keras as hvd_keras
hvd_keras.allreduce([0], name="Barrier")
```
But I can't figure out an equivalent to work with PyTorch? Any suggestion for an idiomatic way to do a Barrier when using Horovod with PyTorch. Thank you.

**Comment by alsrgv** [comment]
@avolkov1, how about this:
```python
import horovod.torch as hvd
hvd.allreduce(torch.tensor(0), name='barrier')
```

**Comment by avolkov1** [comment]
Thank you! That worked great. I'm not as familiar with PyTorch. I was trying `hvd.allreduce(torch.FloatTensor(0), name='barrier')` and that didn't seem to work before. Although now that also seemed to work. Thank you again.

**Comment by UditGupta10** [comment]
@alsrgv 
> @ppwwyyxx, `hvd.allreduce([], name="Barrier")` would produce an equivalent effect to an MPI_Barrier. But it would still print a warning that other nodes are waiting on the node running the validation, which can be ignored. Would that work for you? We can think about finer control over warnings if you don't want to see them.

I am also trying to use a barrier in tensorflow. I suppose hvd refers to import horovod.tensorflow as hvd.
Upon looking the documentation, the allreduce has no argument 'name'. What is the standard way to use a barrier in code as of now?

**Comment by ppwwyyxx** [comment]
I'm using `barrier = hvd.allreduce(tf.random_normal(shape=[1]))`.

**Comment by UditGupta10** [comment]
Again, what is hvd?
Is 'barrier' a dummy returned value that has no significance?

**Comment by ppwwyyxx** [comment]
> So basically, whenever an mpi barrier is required we have to do sess.run(barrier) ?

Roughly yes.
`sess.run(tensor)` is not the only way a tensor can be evaluated. You can also `sess.run(other_tensor)` if `other_tensor` depends on `tensor`.

**Comment by UditGupta10** [comment]
Thanks for the prompt replies..

**Comment by swethmandava** [comment]
I have a similar issue. I tried using all reduce like this

`
if master_process:
    //code
barrier = hvd.allreduce(tf.constant(0))
with tf.Session(config=config) as sess:
    sess.run(barrier)
estimator.train(.....)
`

But this gives me missing_rank:0 for as long as it takes //code to run and then missing_rank:[1,2,3,4,5,6,7] after that and just hangs. Can I increase the 60s limit? The //code is preparing some files and takes time (I'm looking into how i can distribute it but nevertheless)

**Comment by UditGupta10** [comment]
I don't think allreduce should be used within if block of rank 0. It should be seen by all ranks so remove the 'if' part, see if it works.

**Comment by swethmandava** [comment]
All reduce is not within the if block. Only //code is.

**Comment by UditGupta10** [comment]
I missed that. Maybe you could use hvd.shutdown before if block and hvd.init after if block again. I haven't used it personally, just give it a try.

**Comment by UditGupta10** [comment]
There's also a naive approach. Initialise a boolean flag to be False for all ranks. In the //code last line set the flag to be True and broadcast using mpi4py. All other ranks check for the flag to be True using while loop, break when true. This will act as a barrier, allreduce (0) not required. Let me know if it helps.

**Comment by abditag2** [comment]
@swethmandava i dont think you can allreduce a constant tensor. Try the barrier with a Variable tensor.

**Comment by alsrgv** [comment]
@swethmandava, we've recently added an option to set stall check time `HOROVOD_STALL_CHECK_TIME_SECONDS`, it will be available in 0.16.2.  You can also get it if you install Horovod from master.
